myapp/views.py

```diff
@@ -59,85 +60,85 @@ def single_stock_historic(request, symbol):
-                                               f'\n your current balance is: {user_profile.balance}'})
-
-                else:
-                    content.update({"error": "not enough balance"})
-                    return render(request, 'trade.html', content)
-
-            elif 'sell' in params:
-                available_stocks = models.UserProfile.objects.get(user=request.user).stocks[stock_symbol]
-                if available_stocks >= number_of_stocks:
-                    t = models.Transaction.objects.create(user_id=models.UserProfile.objects.get(user=request.user),
-                                                          stock_symbol=stock_symbol,
-                                                          trans_date=datetime.now(),
-                                                          buy_or_sell=1,
-                                                          quantity=number_of_stocks,
-                                                          price=price)
-                    t.save()
-
-                    user_profile.stocks[stock_symbol] -= number_of_stocks
-
-                    user_profile.save()
-
-                    content.update({'success': f'You have sold {number_of_stocks} for a price {price}'
-                                               f'\n your current balance is: {user_profile.balance}'})
-                else:
-                    content.update({"error": "not enough stocks"})
-        else:
-            return redirect('login')
+    content = {'stock_list': stock_list}
+    rendered_page = render(request, 'trade.html', content)
+    return rendered_page
 
 
+@login_required
+@csrf_exempt
+def buy(request):
+    stock_list = stock_api.get_all_stocks()
+    context = {'stock_list': stock_list}
+    if request.method == 'POST':
+        params = request.POST
 
+        stock_info = trade_logic.extract_info(params)
 
-    rendered_page = render(request, 'trade.html', content)
-    return rendered_page
+        if stock_info is None:
+            context.update({"error": "Invalid input"})
+            return render(request, 'trade.html', context)
 
+        user_profile = models.UserProfile.objects.get(user=request.user)
+
+        if stock_info['total_price'] > user_profile.balance:
+            context.update({"error": "not enough balance"})
+            return render(request, 'trade.html', context)
+
+        user_profile.balance -= stock_info['total_price']
+        user_profile.save()
+
+        trade_logic.create_transaction(request.user, stock_info)
+
+        context.update({'success': f'You have bought {stock_info["number_of_stocks"]} {stock_info["symbol"]} '
+                                   f'for a price {stock_info["total_price"]}'
+                                   f'\n your current balance is: {user_profile.balance}'})
+
+        return render(request, 'trade.html', context)
+
+    return render(request, 'trade.html', context)
+
+@login_required
+@csrf_exempt
+def sell(request):

```

Comment:
the only difference between but and sell is the logic of if this transtaction is allowed. 
See if you can remove duplicate code

Reply:
That will make a long view: 50 lines or so